pkg/core/bundle_list.go

```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
-	settings := CoreSettings{concurrentBundleList: defaultBundleListConcurrency}
+// Example usage: printing bundle descriptors as they come
+//
+//   err := core.ListBundlesApply(repo, store, func(bundle model.BundleDescriptorOption) error {
+//				fmt.Fprintf(os.Stderr, "%v\n", bundle)
+//				return nil
+//			})
+func ListBundlesApply(repo string, store storage.Store, apply ApplyBundleFunc, opts ...BundleListOption) error {
+	var (
+		err, applyErr error
+		wg            sync.WaitGroup
+		workDone      bool
+		mutex         sync.Mutex
+	)
+
+	bundleChan := make(chan model.BundleDescriptor)
+	doneChan := make(chan struct{}, 1)
+
+	// collect bundle metadata asynchronously
+	wg.Add(1)
+	go func(bundleChan chan<- model.BundleDescriptor, doneChan chan struct{}, wg *sync.WaitGroup) {
+		defer func() {
+			close(bundleChan)
+			wg.Done()
+		}()
+
+		bundlesChan, errChan, workers := ListBundlesChan(repo, store, append(opts, WithBundleDoneChan(doneChan))...)
+
+		err = DoSelectBundles(bundlesChan, errChan, func(bundleBatch model.BundleDescriptors) {
+			for _, bundle := range bundleBatch {
+				bundleChan <- bundle // transfer a batch of metadata to the applied func
+			}
+		})
+
+		mutex.Lock()
+		workDone = true // nothing left to interrupt
+		close(doneChan)
+		mutex.Unlock()
+
+		workers.Wait()
+	}(bundleChan, doneChan, &wg)
+
+	// apply function on collected metadata
+	for bundle := range bundleChan {
+		if applyErr = apply(bundle); applyErr != nil {
+			// wind down goroutines, but if interrupt channel is already closed (i.e nothing left to be interrupted)
+			mutex.Lock()
+			if !workDone {
+				doneChan <- struct{}{}
+			}
+			mutex.Unlock()
+			for range bundleChan {
+			} // wait for close
+			break
+		}
+	}
+	wg.Wait()
+	// collect errors
+	switch {
+	case err == ErrInterrupted && applyErr != nil:
+		return applyErr
+	case err != nil:
+		return err
+	case applyErr != nil:
+		return applyErr
+	default:
+		return nil
+	}
+}
+
+// ListBundles returns a list of bundle descriptors from a repo. It collects all bundles until completion.
+func ListBundles(repo string, store storage.Store, opts ...BundleListOption) (model.BundleDescriptors, error) {
+	bundles := make(model.BundleDescriptors, 0, typicalBundlesNum)
+
+	bundlesChan, errChan, workers := ListBundlesChan(repo, store, opts...)
+
+	// consume batches of ordered bundles
+	err := DoSelectBundles(bundlesChan, errChan, func(bundleBatch model.BundleDescriptors) {
+		bundles = append(bundles, bundleBatch...)
+	})
+
+	workers.Wait()
+
+	return bundles, err // we may have some batches resolved before the error occurred
+}
+
+// ListBundlesChan returns a list of bundle descriptors from a repo. Each batch of returned descriptors
+// is sent on the output channel, following key lexicographic order.
+//
+// Simple use cases of this helper are wrapped in ListBundles (block until completion) and ListBundlesApply
```

Comment:
let's remove the cruft:  if there's no dependency chain of `ListBundles` percolating to the ui, there's no need for `ListBundles`

Reply:
.. although it could be reasonable to add another commit to the mainline for the sake of having the bifurcated (sync/async) functionality in history

Reply:
Used by web utility